[Bug] tcgdiskstat - `runtime error: index out of range [3] with length 1`

I can observe a panic when using `tcgdiskstat` on SCSI devices (/dev/sdX). That mostly happens with USB-Sticks or SATA-to-USB adapter that are connected to the PC, when the command is executed.

This is an inconvenience, as the devices can be unplugged and my tested devices did not support SSC anyhow.

However, investigation into the panic shows that this is triggered by sub-optimal code inside the `sgio.SCSIInquiry()`function. That handles the responses of the device messages.

**Error Message**
```
sudo <HOME>/go/bin/tcgdiskstat
panic: runtime error: index out of range [3] with length 1

goroutine 1 [running]:
github.com/open-source-firmware/go-tcg-storage/pkg/drive/sgio.SCSIInquiry(0x6)
        <HOME>/go/pkg/mod/github.com/open-source-firmware/go-tcg-storage@v0.8.22-0.20230920111945-87417253ae2d/pkg/drive/sgio/ops.go:177 +0x64b
github.com/open-source-firmware/go-tcg-storage/pkg/drive.isSCSI({0x8de298?, 0xc000062258?})
        <HOME>/go/pkg/mod/github.com/open-source-firmware/go-tcg-storage@v0.8.22-0.20230920111945-87417253ae2d/pkg/drive/scsi_nix.go:90 +0x27
github.com/open-source-firmware/go-tcg-storage/pkg/drive.Open({0xc0000135c0?, 0xc0000f9d90?})
        <HOME>/go/pkg/mod/github.com/open-source-firmware/go-tcg-storage@v0.8.22-0.20230920111945-87417253ae2d/pkg/drive/drive_nix.go:19 +0x65
github.com/open-source-firmware/go-tcg-storage/pkg/core.NewCore({0xc0000135c0, 0x8})
        <HOME>/go/pkg/mod/github.com/open-source-firmware/go-tcg-storage@v0.8.22-0.20230920111945-87417253ae2d/pkg/core/dev.go:22 +0x27
main.main()
        <HOME>/go/pkg/mod/github.com/open-source-firmware/go-tcg-storage@v0.8.22-0.20230920111945-87417253ae2d/cmd/tcgdiskstat/main.go:65 +0x3e8

```

* In case of one device (Sharkoon branded USB-Stick) the device would answer to the VPD request without raising an error but the message was wrong/incomplete. Leading to the shown panic inside the device information querry.
* Another device - can't find it anymore - would already panic when parsing the serial number response.

I see two main problems here:
1. The code should generally be more defensive when accessing the returned data and check bounds and expected values. Eg. the length and the page code of the response.
2. To be extra careful the VPD page 0x00 should be queried to ensure the relevant pages are supported/implemented by the device.

Note: That page itself may not be present in very old devices. In that case the code should probably fail/stop gracefully.

--- a/pkg/drive/sgio/ops.go
+++ b/pkg/drive/sgio/ops.go
@@ -162,7 +162,7 @@ func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 	/* fixup length field to indicate full page length */
 	l := inqHdr.Length + 5
 	if l < SCSI_INQUIRY_STD_LENGTH {
-		return nil, fmt.Errorf("unexpected respsonse length of SCSI_INQUIRY (%u < %u)", l, SCSI_INQUIRY_STD_LENGTH)
+		return nil, fmt.Errorf("unexpected respsonse length of SCSI_INQUIRY (%d < %d)", l, SCSI_INQUIRY_STD_LENGTH)
 	}
 
 	respBuf = make([]byte, SCSI_VPD_STD_LENGTH)

--- a/pkg/drive/sgio/ops.go
+++ b/pkg/drive/sgio/ops.go
@@ -18,6 +18,7 @@ package sgio
 import (
 	"bytes"
 	"encoding/binary"
+	"errors"
 	"fmt"
 	"strings"
 )
@@ -37,10 +38,14 @@ const (
 	SCSI_INQUIRY            = 0x12
 	SCSI_INQUIRY_STD_LENGTH = 0x24 // expected minimal length of SCSI_INQUERY according to SPC-3 (and newer)
 
-	SCSI_VPD_STD_LENGTH = 0xFF // max page size - should be enoough for most VPDs
+	SCSI_VPD_STD_LENGTH = 0xFF // max page size - should be enough for most VPDs
 	SCSI_VPD_PAGE_SV    = 0x00 // VPD page indicating other supported VPD pages
 	SCSI_VPD_PAGE_SN    = 0x80 // Unit serial number VPD page
 	SCSI_VPD_PAGE_DI    = 0x83 // Device Identification VPD page
+)
+
+var (
+	ErrUnexpectedResp = errors.New("unexpected SCSI response")
 )
 
 type SCSIProtocol int
@@ -127,13 +132,11 @@ func (id IdentifyDeviceResponse) String() string {
 		strings.TrimSpace(ATAString(id.Model[:])))
 }
 
-/*
-	 INQUIRY - Returns parsed inquiry data.
-		- request standard inquiry first
-		- check supported VPDs
-		- query for serial number, if page 0x80 is supported
-	    - query for protocol type, if page 0x83 is supported
-*/
+//	 INQUIRY - Returns parsed inquiry data.
+//		- request standard inquiry first
+//		- check supported VPDs
+//		- query for serial number, if page 0x80 is supported
+//	    - query for protocol type, if page 0x83 is supported
 func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 	respBuf := make([]byte, SCSI_INQUIRY_STD_LENGTH)
 
@@ -159,10 +162,10 @@ func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 		return nil, err
 	}
 
-	/* fixup length field to indicate full page length */
+	// fixup length field to indicate full page length
 	l := inqHdr.Length + 5
 	if l < SCSI_INQUIRY_STD_LENGTH {
-		return nil, fmt.Errorf("unexpected respsonse length of SCSI_INQUIRY (%d < %d)", l, SCSI_INQUIRY_STD_LENGTH)
+		return nil, fmt.Errorf("%w: length of SCSI_INQUIRY (%d < %d)", ErrUnexpectedResp, l, SCSI_INQUIRY_STD_LENGTH)
 	}
 
 	respBuf = make([]byte, SCSI_VPD_STD_LENGTH)
@@ -181,7 +184,7 @@ func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 	haveSN := false
 	haveDI := false
 
-	/* validate response */
+	// validate response
 	l = vpdHdr.PageLength + 4
 	if (vpdHdr.PageCode == SCSI_VPD_PAGE_SV) && (l > 4) && (l <= SCSI_VPD_STD_LENGTH) {
 		supList := respBuf[4:l]
@@ -228,13 +231,12 @@ func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 
 		didlen := binary.BigEndian.Uint16(respBuf[2:4]) + 4 // page length (n-3)
 		if (respBuf[1] == SCSI_VPD_PAGE_DI) && (didlen > 4) && (didlen <= 2048) {
-			/* Device Identification VPD page - decode length and descriptor list */
+			// Device Identification VPD page - decode length and descriptor list
 
 			did := respBuf[4:didlen] // identification descriptor list (full)
 
-			/* We are only interested in the protocol identifier.
-			   Loop through all ID descriptors and check for a valid protocol field.
-			*/
+			// We are only interested in the protocol identifier.
+			// Loop through all ID descriptors and check for a valid protocol field.
 			for {
 				if len(did) <= 4 {
 					break

--- a/pkg/drive/sgio/ops.go
+++ b/pkg/drive/sgio/ops.go
@@ -28,12 +28,19 @@ const (
 	ATA_TRUSTED_SND     = 0x5e
 	ATA_IDENTIFY_DEVICE = 0xec
 
-	SCSI_INQUIRY          = 0x12
 	SCSI_MODE_SENSE_6     = 0x1a
 	SCSI_READ_CAPACITY_10 = 0x25
 	SCSI_ATA_PASSTHRU_16  = 0x85
 	SCSI_SECURITY_IN      = 0xa2
 	SCSI_SECURITY_OUT     = 0xb5
+
+	SCSI_INQUIRY            = 0x12
+	SCSI_INQUIRY_STD_LENGTH = 0x24 // expected minimal length of SCSI_INQUERY according to SPC-3 (and newer)
+
+	SCSI_VPD_STD_LENGTH = 0xFF // max page size - should be enoough for most VPDs
+	SCSI_VPD_PAGE_SV    = 0x00 // VPD page indicating other supported VPD pages
+	SCSI_VPD_PAGE_SN    = 0x80 // Unit serial number VPD page
+	SCSI_VPD_PAGE_DI    = 0x83 // Device Identification VPD page
 )
 
 type SCSIProtocol int
@@ -78,6 +85,13 @@ type InquiryResponse struct {
 	SerialNumber []byte
 }
 
+type SimpleVPDResponse struct {
+	Peripheral byte
+	PageCode   byte
+	_          byte
+	PageLength byte
+}
+
 func (inq InquiryResponse) String() string {
 	return fmt.Sprintf("Type=0x%x, Vendor=%s, Product=%s, Serial=%s, Revision=%s",
 		inq.Peripheral,
@@ -113,11 +127,17 @@ func (id IdentifyDeviceResponse) String() string {
 		strings.TrimSpace(ATAString(id.Model[:])))
 }
 
-// INQUIRY - Returns parsed inquiry data.
+/*
+	 INQUIRY - Returns parsed inquiry data.
+		- request standard inquiry first
+		- check supported VPDs
+		- query for serial number, if page 0x80 is supported
+	    - query for protocol type, if page 0x83 is supported
+*/
 func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
-	respBuf := make([]byte, 36)
+	respBuf := make([]byte, SCSI_INQUIRY_STD_LENGTH)
 
-	cdb := CDB6{SCSI_INQUIRY}
+	cdb := CDB6{SCSI_INQUIRY, 0} /* no VPD */
 	binary.BigEndian.PutUint16(cdb[3:], uint16(len(respBuf)))
 
 	if err := SendCDB(fd, cdb[:], CDBFromDevice, &respBuf); err != nil {
@@ -128,59 +148,108 @@ func SCSIInquiry(fd uintptr) (*InquiryResponse, error) {
 		Peripheral   byte // peripheral qualifier, device type
 		_            byte
 		Version      byte
-		_            [5]byte
+		_            byte
+		Length       byte
+		_            [3]byte
 		VendorIdent  [8]byte
 		ProductIdent [16]byte
 		ProductRev   [4]byte
 	}{}
-
 	if err := binary.Read(bytes.NewBuffer(respBuf), nativeEndian, &inqHdr); err != nil {
 		return nil, err
 	}
 
-	respBuf = make([]byte, 128)
-	cdb = CDB6{SCSI_INQUIRY}
-	cdb[1] = 0x1 /* Request VPD page 0x80 for serial number */
-	cdb[2] = 0x80
+	/* fixup length field to indicate full page length */
+	l := inqHdr.Length + 5
+	if l < SCSI_INQUIRY_STD_LENGTH {
+		return nil, fmt.Errorf("unexpected respsonse length of SCSI_INQUIRY (%u < %u)", l, SCSI_INQUIRY_STD_LENGTH)
+	}
+
+	respBuf = make([]byte, SCSI_VPD_STD_LENGTH)
+	cdb = CDB6{SCSI_INQUIRY, 0x1, SCSI_VPD_PAGE_SV} // Request VPD page 0x00 - supported VPDs
 	binary.BigEndian.PutUint16(cdb[3:], uint16(len(respBuf)))
 
 	if err := SendCDB(fd, cdb[:], CDBFromDevice, &respBuf); err != nil {
 		return nil, err
 	}
 
-	snHdr := struct {
-		_      [3]byte
-		Length byte
-	}{}
-	if err := binary.Read(bytes.NewBuffer(respBuf), nativeEndian, &snHdr); err != nil {
-		return nil, err
-	}
-	sn := respBuf[4 : 4+snHdr.Length]
-
-	respBuf = make([]byte, 2048)
-	cdb = CDB6{SCSI_INQUIRY}
-	cdb[1] = 0x1 /* Request VPD page 0x83 for device ID */
-	cdb[2] = 0x83
-	binary.BigEndian.PutUint16(cdb[3:], uint16(len(respBuf)))
-
-	if err := SendCDB(fd, cdb[:], CDBFromDevice, &respBuf); err != nil {
+	var vpdHdr SimpleVPDResponse
+	if err := binary.Read(bytes.NewBuffer(respBuf), nativeEndian, &vpdHdr); err != nil {
 		return nil, err
 	}
 
-	didlen := binary.BigEndian.Uint16(respBuf[2:4])
-	did := respBuf[4 : didlen+4]
+	haveSN := false
+	haveDI := false
+
+	/* validate response */
+	l = vpdHdr.PageLength + 4
+	if (vpdHdr.PageCode == SCSI_VPD_PAGE_SV) && (l > 4) && (l <= SCSI_VPD_STD_LENGTH) {
+		supList := respBuf[4:l]
+
+		for i := range supList {
+			if supList[i] == SCSI_VPD_PAGE_SN {
+				haveSN = true
+			}
+			if supList[i] == SCSI_VPD_PAGE_DI {
+				haveDI = true
+			}
+		}
+	}
+
+	sn := []byte(nil)
+	if haveSN {
+		respBuf = make([]byte, SCSI_VPD_STD_LENGTH)
+		cdb = CDB6{SCSI_INQUIRY, 0x1, 0x80} // Request VPD page 0x80 - serial number
+		binary.BigEndian.PutUint16(cdb[3:], uint16(len(respBuf)))
+
+		if err := SendCDB(fd, cdb[:], CDBFromDevice, &respBuf); err != nil {
+			return nil, err
+		}
+
+		if err := binary.Read(bytes.NewBuffer(respBuf), nativeEndian, &vpdHdr); err != nil {
+			return nil, err
+		}
+
+		l := vpdHdr.PageLength // sn page length includes header!
+		if (vpdHdr.PageCode == SCSI_VPD_PAGE_SN) && (l > 4) && (l <= SCSI_VPD_STD_LENGTH) {
+			sn = respBuf[4:l]
+		}
+	}
+
 	proto := SCSIProtocol(-1)
-	for {
-		if len(did) == 0 {
-			break
+	if haveDI {
+		respBuf = make([]byte, 2048)
+		cdb = CDB6{SCSI_INQUIRY, 0x1, 0x83} // Request VPD page 0x80 - device identification
+		binary.BigEndian.PutUint16(cdb[3:], uint16(len(respBuf)))
+
+		if err := SendCDB(fd, cdb[:], CDBFromDevice, &respBuf); err != nil {
+			return nil, err
 		}
-		l := did[3]
-		part := did[:l+4]
-		piv := part[1]&0x80 > 0
-		if piv {
-			proto = SCSIProtocol(part[0] >> 4)
+
+		didlen := binary.BigEndian.Uint16(respBuf[2:4]) + 4 // page length (n-3)
+		if (respBuf[1] == SCSI_VPD_PAGE_DI) && (didlen > 4) && (didlen <= 2048) {
+			/* Device Identification VPD page - decode length and descriptor list */
+
+			did := respBuf[4:didlen] // identification descriptor list (full)
+
+			/* We are only interested in the protocol identifier.
+			   Loop through all ID descriptors and check for a valid protocol field.
+			*/
+			for {
+				if len(did) <= 4 {
+					break
+				}
+
+				l := did[3]                    // identifier length (n-3)
+				part := did[:l+4]              // identifier descriptor
+				piv := (part[1] & 0x80) > 0    // protocol identifier valid bit
+				assoc := (part[1] & 0x30) >> 4 // association field
+				if piv && assoc >= 1 && assoc <= 2 {
+					proto = SCSIProtocol(part[0] >> 4)
+				}
+				did = did[l+4:]
+			}
 		}
-		did = did[l+4:]
 	}
 	resp := InquiryResponse{
 		Protocol:     proto,

